server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedSortedSetDocValuesSyntheticFieldLoader.java

```diff
@@ -13,27 +13,43 @@
 import org.apache.lucene.index.LeafReader;
 import org.apache.lucene.index.SortedSetDocValues;
 import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.core.Nullable;
 import org.elasticsearch.index.mapper.SourceLoader;
 import org.elasticsearch.xcontent.XContentBuilder;
 
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeSet;
+import java.util.stream.Stream;
 
-public class FlattenedSortedSetDocValuesSyntheticFieldLoader extends SourceLoader.DocValuesBasedSyntheticFieldLoader {
-    private DocValuesFieldValues docValues = NO_VALUES;
+public class FlattenedSortedSetDocValuesSyntheticFieldLoader implements SourceLoader.SyntheticFieldLoader {
     private final String fieldFullPath;
     private final String keyedFieldFullPath;
+    private final String keyedIgnoredValuesFieldFullPath;
     private final String leafName;
 
+    private DocValuesFieldValues docValues = NO_VALUES;
```

Comment:
Can we just use `SortedSetDocValues` here directly instead of using an additional layer of abstraction? 

For example in `DocValuesSortedKeyedValues` we directly sorted set doc values instead of `DocValuesFieldValues` as an intermediate interface.

Reply:
Do we need the `DocValuesFieldValues` interface at all? It is either no value or uses `SortedSetDocValues`. Maybe we always use `SortedSetDocValues` directly and in case of no value it is null?

Reply:
I was actually looking at that too. I believe we can't directly use `SortedSetDocValues` because when there is no doc values, it contains an empty singleton which we don't expect. We could just set it to null indeed.

Reply:
So `FlattenedFieldDocValuesLoader` also tracks if current document has doc values and i think inlining all this logic into `FlattenedSortedSetDocValuesSyntheticFieldLoader` reduces readability quite a bit. I'll keep this as is.